plugins/Diagnostics/Diagnostic/PrivateDirectories.php

```diff
@@ -0,0 +1,155 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Plugins\Diagnostics\Diagnostic;
+
+use Piwik\Common;
+use Piwik\Filesystem;
+use Piwik\Http;
+use Piwik\SettingsPiwik;
+use Piwik\Translation\Translator;
+
+abstract class PrivateDirectories implements Diagnostic
+{
+    protected $privatePaths = [];
+
+    protected $labelKey = 'Diagnostics_RequiredPrivateDirectories';
+
+    /**
+     * @var Translator
+     */
+    protected $translator;
+
+    public function __construct(Translator $translator)
+    {
+        $this->translator = $translator;
+    }
+
+    public function execute()
+    {
+        if (!SettingsPiwik::isMatomoInstalled()) {
+            return [];
+        }
+
+        $label = $this->translator->translate($this->labelKey);
+
+        $privatePaths = $this->privatePaths;
+
+        // create test file to check if tmp/empty exists Note: This won't work in a load balanced environment!
+        Filesystem::mkdir(PIWIK_INCLUDE_PATH . '/tmp');
+        file_put_contents(PIWIK_INCLUDE_PATH . '/tmp/empty', 'test');
+
+        $baseUrl = SettingsPiwik::getPiwikUrl();
+        if (!Common::stringEndsWith($baseUrl, '/')) {
+            $baseUrl .= '/';
+        }
+
+        $manualCheck = $this->translator->translate('Diagnostics_PrivateDirectoryManualCheck');
+
+        $testUrls = [];
+        foreach ($privatePaths as $checks) {
+            foreach ($checks as $path) {
+                if (!file_exists($path)) {
+                    continue;
+                }
+
+                $testUrls[] = $baseUrl . $path;
+            }
+        }
+
+        $isInternetEnabled = SettingsPiwik::isInternetEnabled();
+        if (!$isInternetEnabled) {
+            $testUrlsList = $this->getUrlList($testUrls);
+
+            $unknown = $this->translator->translate('Diagnostics_PrivateDirectoryInternetDisabled') . ' ' . $manualCheck
+                . $testUrlsList;
+            $results[] = DiagnosticResult::singleResult($label, DiagnosticResult::STATUS_WARNING, $unknown);
+            return $results;
+        }
+
+        $result = new DiagnosticResult($label);
+
+        $isConfigIniAccessible = $this->isAccessible($result, $baseUrl . 'config/config.ini.php', ';', 'trusted_hosts[]');
```

Comment:
Can we execute this check only for the required directory check to speed things up etc?